varinfo/variable.py

```diff
@@ -341,17 +341,51 @@ class VariableFromDmr(VariableBase, AttributeContainerFromDmr):
 
     """
 
+    def __init__(
+        self,
+        element,
+        cf_config,
+        namespace,
+        full_name_path,
+        dim_name_size=[str, int],
+    ):
+        self.dim_name_size = dim_name_size
+        super().__init__(element, cf_config, namespace, full_name_path)
+
     def _get_data_type(self, variable: ET.Element) -> str:
         """Extract a string representation of the variable data type."""
         return variable.tag.lstrip(self.namespace).lower()
 
     def _get_shape(self, variable: ET.Element) -> tuple[int]:
-        """Extract the shape of the variable data array. This is not yet
-        implemented as the Dimension information is currently unavailable
-        to the Variable XML content.
-
+        """Extract the shape of the variable data array. First explore
+        XML for - DIM child elements with a size
+        attribute - true for HDF5 files and anonymous dimensions
+        (note no name). If found (with or without names), use them. If
+        not found, use dim_name_size dictionary with matching dimension
+        names to define the variable shape. Note the dim_name_size
+        dictionary is filled when parsing xml for variables, capturing
+        dimensions names and sizes for dimension entries - as found in
+        NetCDF files with named dimensions.
         """
 
+        # Retrieve the shape from the dimension size in the
+        # <Dim size=xx/> element.
+        self.shape = [
+            int(dim_size.attrib['size'])
+            for dim_size in variable.findall(f'.//{self.namespace}Dim[@size]')
+        ]
+
+        # Retrieve the shape from dimension size in the
+        # <Dimension size=xx/> element
+        if not self.shape:
+            self.shape = [
+                self.dim_name_size[dim_name]
+                for dim_name in self.dimensions
+                if dim_name in self.dim_name_size
+            ]
+
+        return self.shape
```

Comment:
General note (that might be OBE depending on if you update this method to always grab sizes from `<Dimension />` elements):

This method is returning the value for `shape` that is assigned to `self.shape` in the `VariableBase.__init__`. Generally, you don't want to both assign to a class attribute within a method, and then return the value to be assigned elsewhere. In this example, it would be better to make a locally scoped variable (maybe `shape`) and then let the assignment happen in `VariableBase.__init__`.

Reply:
Marking as resolved. Thanks for addressing this.